First draft of "No More Click-Ops" blog post (#12259)
pulumi/docs: content/blog/next-level-iac-no-more-clickops/index.md

---
title: "Next-level IaC: ClickOps No More"

date: 2024-07-15
draft: false
social_media: "Many organizations start out using Click-Ops, but find themselves digging a hole they can't get back out of. Pulumi to the rescue!"
meta_desc: "How to use Pulumi to dig yourself out of a Click-Ops hole."
meta_image: meta.png
authors:
    - troy-howard
tags:
    - next-level-iac
    - python
    - clickops
    - boto
    - import
---

Many organizations start out using Click-Ops, but find themselves digging a hole they can't find their way out of. Today we're going to show you how to use Pulumi to find your way out.

<!--more-->

> Imagine this; you, a valiant keyboard warrior, skilled in the dark arts of Python, just joined a new company to do DevOps... and inherited some existing cloud infrastructure. The person before you just clicked around in AWS making things, instead of using infrastructure-as-code to manage it. Ugh... We call that "Click-Ops". Congratulations, you’ve found yourself in a Click-Ops hole.

Reviewer: I am not sure semicolon is the correct punctuation mark after "Imagine this"

Author: You're right! It should be a colon.

> The biggest problem with Click-Ops is reproducibility. There is no easy way to reliably recreate your environment. Suppose you want to be able to create a staging environment to test some changes, and you also want to be able to recreate the production environment in case of disaster. With Click-Ops that's a serious pain.

Reviewer: You might add one sentence that says, "this is the strong suite of IaC because you are able to reliably reproduce infrastructure changes without manual errors" or something like that

> The biggest problem with Click-Ops is reproducibility. There is no easy way to reliably recreate your environment. Suppose you want to be able to create a staging environment to test some changes, and you also want to be able to recreate the production environment in case of disaster. With Click-Ops that's a serious pain.

Reviewer 2: ```suggestion
The biggest problem with Click-Ops is reproducibility. There is no easy way to reliably recreate your environment. Suppose you want to be able to create a staging environment to test some changes, and you also want to be able to recreate the production environment in case of disaster. With Click-Ops, that's a serious pain.
```

Let's fix that... by bringing our existing infrastructure under management by Pulumi.

## Explore your existing AWS infrastructure

The first step to undoing a Click-Ops mess is to have a look at what is already there. You can do this via the AWS web-based experience, or if you're in love with the command-line like we are, we can use the AWS CLI.

> In this case we're going to assume that there is a existing AWS VPC that we want to bring under management by Pulumi. There are likely more than one, so lets look at a list of VPCs:

Reviewer 2: ```suggestion
In this case we're going to assume that there is a existing AWS VPC that we want to bring under management by Pulumi. There is likely more than one, so lets look at a list of VPCs:
```

```shell
$ aws ec2 describe-vpcs --output JSON
```

> This command shows you all the VPCs under your AWS account, in JSON format. Pulumi will let us describe and manage all of that with code, but creating it all from scratch that would need a lot of typing. Luckily, we can import this existing infrastructure with a single command `pulumi import`, which takes a list of resources in JSON format and outputs a Pulumi program in your preferred programming language.

Reviewer 2: ```suggestion
This command shows you all the VPCs under your AWS account, in JSON format. Pulumi will let us describe and manage all of that with code, but creating it all from scratch would mean a lot of manual typing. Luckily, we can import this existing infrastructure with a single command `pulumi import`, which takes a list of resources in JSON format and outputs a Pulumi program in your preferred programming language.
```

## Using Boto to generate Pulumi-compatible JSON

{{% notes type="info" %}}
The command-line tool `jq` is really handy for viewing and filtering JSON. If you don't already have it set up go ahead and [get that installed](https://jqlang.github.io/jq/download/).
{{% /notes %}}

We put together a basic demo tool called [`pulumi-import-aws-account-scraper`][pulumi-import-aws-account-scraper] using Python’s [Boto library][boto-lib], which can scrape your AWS account, finding all the various resources. It outputs Pulumi compatible JSON which we can then use to generate a Pulumi program in your preferred language.

First, clone the [`pulumi/pulumi-import-aws-account-scraper`][pulumi-import-aws-account-scraper] repo and setup the tool:

```shell
git clone git@github.com:pulumi/pulumi-import-aws-account-scraper.git
cd pulumi-import-aws-account-scraper
virtualenv -p python3 venv
source venv/bin/activate
pip install -r requirements.txt
```

Next we will use this demo tool to scan your account:

```shell
source venv/bin/activate
python3 account_scraper.py > resources.json
```

> This will generate a JSON file with an entry for each resource under your account. However, we only want to import some of those. For this example, we'll assume there is a VPC named `thor` in that list, with associated resources also containing the term `thor` in their name. We can filter that out with a quick `jq` expression:

Reviewer: Does it make sense to show the JSON outputted at all?

Also is it useful to explain the format at all?

Reviewer 2: I agree here about showing the JSON output.

```shell
cat resources.json | jq '.resources |= map(select(.name | contains("thor")))' > resources-filtered.json
```

If you're working with a different resource type, with different names, feel free to customize this expression to filter out anything you don't want to bring under management by Pulumi.

## Run pulumi import to generate a Pulumi program

Ok, now we've got a file called `resources-filtered.json` in Pulumi-compatible JSON format, describing only the resources we want to import for our particular VPC.

> Here's the magic moment! Let’s import that into Pulumi and generate our Pulumi program!

Reviewer: I think we should double click here and explain what is happening. It is importing the existing VPC into the state store and then generating code for it. Now this state piece if important because if you just execute the code, well it will try to create anew.

Reviewer 2: I also agree here. When I was reading this, I was wondering exactly what was happening that was letting Pulumi know to manage existing resources rather than just create new ones since nothing in the IaC code was visibly indicating that an import was happening.

## Next Steps

If you haven't already, [install Pulumi][pulumi-install] today, and follow our self-directed [Getting Started guides][pulumi-start] to learn more about making the most of Pulumi's next-level infrastructure management features at your organization.

To learn more, you can watch the following video which provides a high level overview of how Pulumi works:

<div class="rounded-md shadow border border-gray-300 w-3/4 mx-auto my-4" style="position: relative; padding-bottom: 40.25%; height: 0; overflow: hidden;">
    <iframe
        src="//www.youtube.com/embed/Q8tw6YTD3ac?rel=0"
        style="position: absolute; top: 0; left: 0; width: 100%; height: 100%; border:0;"
        allowfullscreen=""
        title="Introduction to Pulumi in Three Minutes"
        srcdoc="<style>*{padding:0;margin:0;overflow:hidden}html,body{height:100%}img{position:absolute;width:100%;top:0;bottom:0;margin:auto}</style><a href=https://www.youtube.com/embed/Q8tw6YTD3ac?autoplay=1><img src='/images/home/youtube-getting-started.png' alt='Introduction to Pulumi in Three Minutes'></a>">
    </iframe>
</div>

## Pulumi Cloud

> The Pulumi Cloud is a fully managed service that helps you adopt Pulumi’s open source SDK with ease. It provides built-in state and secrets management, integrates with source control and CI/CD, and offers a web console and API that make it easier to visualize and manage infrastructure. It is free for individual use, with features available for teams.

Reviewer: Nice touch to add a CTA here. Thanks
